### ISSUE TITLE
Exception handling in Execute method of BusinessRule(s)

### ISSUE THREAD

**Issue by Chicagoan2016** [MarimerLLC/cslaforum#778]
Hi All,
we have business rules to calculate/verify tax numbers. We consume RO collections within those business rules. So far we are calling factory methods of RO collections within Execute method.

Two questions come to mind regarding this.
1) Is calling factory methods on RO collections (essentially to fetch data into these collections) the 'stanadard' practice?
2) How should we go about exception handling in Execute method of Business rules? 

Kind regards

**Version and Platform**
CSLA version: 4.7.100
OS: Windows, Linux, iOS, Android, etc.
Platform: WinForms, WPF, ASP.NET Core, MVC, Xamarin, etc.

**Comment by ajj7060** [comment]
Yes, this is expected if you need more information for the rule to execute; nothing wrong with calling the factory methods for readonly objects from Execute.  (Except for coupling the rule to the concrete RO types).

Just be aware though that if you want your rule to run under something Xamarin, the rule must be async, and you must create & use a factory method which does `DataPortal.FetchAsync`.  Otherwise the device may kill your app for not staying responsive.

IIRC, if you do nothing and an exception occurs in your Execute override, the exception is caught and the rule is marked as broken.  That's probably the most logical way to proceed; if the rule failed to run, the object is in an invalid state (the "brokenness" of the rule is unknown).  Of course you can do your own try/catch and ignore exceptions, it depends on how you want the app to react to failing to run.

**Comment by Chicagoan2016** [comment]
@ajj7060 , today it (Exception) happened in one of my business rules that calls an ROCollection in Execute method.
With no exception handling in Execute method, the error message generated by the rule wasn't helpful, it was something I won't show to the user. The message was essentially details about the RO class and what happened inside the RO class.
I think I will have to come up with some exception handling strategy for my business rules.
With that being said, should csla's business rule engine be modified to accommodate  exceptions in the Execute method? so that users get a default error message?

kind regards
